openmls/src/group/mls_group/proposals.rs

```diff
@@ -205,3 +194,267 @@ impl StagedProposalQueue {
             .map(move |reference| self.get(reference).unwrap())
     }
 }
+
+/// Alternative representation of a Proposal, where the sender is extracted from
+/// the encapsulating MlsPlaintext and the ProposalReference is attached.
+#[derive(Debug, Clone)]
+pub(crate) struct QueuedProposal<'a> {
+    proposal: &'a Proposal,
+    proposal_reference: ProposalReference,
+    sender: Sender,
+    proposal_or_ref_type: ProposalOrRefType,
+}
+
+impl<'a> QueuedProposal<'a> {
+    /// Creates a new [QueuedProposal] from a [Proposal] and [Sender].
+    /// Note that the proposal type will always be `Proposal`.
+    pub(crate) fn from_proposal_and_sender(
+        ciphersuite: &Ciphersuite,
+        backend: &impl OpenMlsCryptoProvider,
+        proposal: &'a Proposal,
+        sender: Sender,
+    ) -> Result<Self, StagedProposalError> {
+        let proposal_reference = ProposalReference::from_proposal(ciphersuite, backend, proposal)?;
+        Ok(Self {
+            proposal,
+            proposal_reference,
+            sender,
+            proposal_or_ref_type: ProposalOrRefType::Proposal,
+        })
+    }
+    /// Creates a new [QueuedProposal] from a [StagedProposal].
+    /// /// Note that the proposal type will always be `Reference`.
+    pub(crate) fn from_staged_proposal(
+        ciphersuite: &Ciphersuite,
+        backend: &impl OpenMlsCryptoProvider,
+        staged_proposal: &'a StagedProposal,
+    ) -> Result<Self, StagedProposalError> {
+        let proposal_reference =
+            ProposalReference::from_proposal(ciphersuite, backend, staged_proposal.proposal())?;
+        Ok(Self {
+            proposal: staged_proposal.proposal(),
+            proposal_reference,
+            sender: *staged_proposal.sender(),
+            proposal_or_ref_type: ProposalOrRefType::Reference,
+        })
+    }
+    /// Returns the `Proposal` as a reference
+    pub(crate) fn proposal(&self) -> &Proposal {
+        self.proposal
+    }
+    /// Returns the `ProposalReference` as a reference
+    pub(crate) fn proposal_reference(&self) -> ProposalReference {
+        self.proposal_reference.clone()
+    }
+    /// Returns the `Sender` as a reference
+    pub(crate) fn sender(&self) -> &Sender {
+        &self.sender
+    }
+}
+
+/// Proposal queue that helps filtering and sorting the Proposals from one
+/// epoch. The Proposals are stored in a `HashMap` which maps Proposal
+/// references to Proposals, such that, given a reference, a proposal can be
+/// accessed efficiently. To enable iteration over the queue in order, the
+/// `ProposalQueue` also contains a vector of `ProposalReference`s.
+#[derive(Default, Debug)]
+pub struct CreationProposalQueue<'a> {
```

Comment:
Why not just call it `ProposalQueue`? Just a nit, but somehow this name sounds weird to me.

Reply:
It used to be called that, but now since there are two queues I wanted to have separate names: `create_commit()` -> `CreationProposalQueue`, `stage_commit_()` -> `StagingProposalQueue`.

Reply:
Fair enough. No strong opinion here.